netlogo-gui/src/main/app/WidgetActions.scala

```diff
@@ -0,0 +1,145 @@
+    bounds
+  }
+
+  def removeSelectionMargin(bounds: Rectangle): Rectangle = {
+    bounds.x += WidgetWrapper.BORDER_E
+    bounds.width -= WidgetWrapper.BORDER_E + WidgetWrapper.BORDER_W
+    bounds.y += WidgetWrapper.BORDER_N
+    bounds.height -= WidgetWrapper.BORDER_N + WidgetWrapper.BORDER_S
+    bounds
+  }
+
+  def moveWidgets(widgetPanel: WidgetPanel): Unit = {
+    val initialMap: Map[WidgetWrapper, Rectangle] = widgetPanel.widgetsBeingDragged.map(a => a -> {
+      addSelectionMargin(a.originalBounds)
+    })(collection.breakOut)
+    val widgets = widgetPanel.widgetsBeingDragged
+    widgetPanel.dropSelectedWidgets()
+    val finalMap: Map[WidgetWrapper, Rectangle] = widgets.map(a => a -> a.getBounds())(collection.breakOut)
+    undoManager.addEdit(new MoveWidgets(widgetPanel, widgets, initialMap, finalMap))
+  }
+
+  def resizeWidget(widgetWrapper: WidgetWrapper): Unit = {
+    val initialBounds = widgetWrapper.originalBounds
+    addSelectionMargin(initialBounds)
+    widgetWrapper.doDrop()
+    undoManager.addEdit(new ResizeWidget(widgetWrapper, initialBounds, widgetWrapper.getBounds()))
+  }
+
+  class AddWidget(widgetPanel: WidgetPanel, widgetWrapper: WidgetWrapper) extends AbstractUndoableEdit {
+    override def redo(): Unit = {
+      widgetPanel.reAddWidget(widgetWrapper)
+    }
+    override def undo(): Unit = {
+      widgetPanel.deleteWidget(widgetWrapper)
+    }
+    override def getPresentationName: String = "Widget Addition"
+  }
+
+  class RemoveWidget(widgetPanel: WidgetPanel, ww: WidgetWrapper) extends AbstractUndoableEdit {
+    override def redo(): Unit = {
+      widgetPanel.deleteWidget(ww)
+    }
+    override def undo(): Unit = {
+      widgetPanel.reAddWidget(ww)
+    }
+    override def getPresentationName: String = "Widget Deletion"
+  }
+
+  class RemoveMultipleWidgets(widgetPanel: WidgetPanel, wws: Seq[WidgetWrapper]) extends AbstractUndoableEdit {
+    override def redo(): Unit = {
+      widgetPanel.deleteWidgets(wws)
+    }
+    override def undo(): Unit = {
+      for(ww <- wws){
+        widgetPanel.reAddWidget(ww)
+      }
+    }
+    override def getPresentationName: String = "Widget(s) Deletion"
+  }
+
+  class MoveWidgets(widgetPanel: WidgetPanel, wws: Seq[WidgetWrapper], initialMap: Map[WidgetWrapper, Rectangle], finalMap: Map[WidgetWrapper, Rectangle]) extends AbstractUndoableEdit {
+    override def redo:Unit = {
+      for(widgetWrapper <- wws){
+        val finalBound = new Rectangle(finalMap(widgetWrapper))
+        if(!widgetWrapper.selected()){
+          removeSelectionMargin(finalBound)
+        }
+        widgetWrapper.setBounds(finalBound)
+      }
+    }
+    override def undo(): Unit = {
+      for(widgetWrapper <- wws){
+        val initialBound = new Rectangle(initialMap(widgetWrapper))
+        if(!widgetWrapper.selected()) {
+          removeSelectionMargin(initialBound)
+        }
+        widgetWrapper.setBounds(initialBound)
+      }
+    }
```

Comment:
`undo` and `redo` are the same method here except for `initialMap`/`finalMap`. Extract a method to avoid repeating code.